pkg/kubeclient/kubelet_client.go

```diff
@@ -158,7 +170,73 @@ func (client *KubeletClient) GetSummary(host string) (*stats.Summary, error) {
 	return summary, err
 }
 
-func (client *KubeletClient) GetMachineInfo(host string) (*cadvisorapi.MachineInfo, error) {
+// get node single-core Frequency, in MHz
+func (client *KubeletClient) GetNodeCpuFrequency(node *v1.Node) (float64, error) {
+	ip, err := util.GetNodeIPForMonitor(node, types.KubeletSource)
+	if err != nil {
+		glog.Errorf("Failed to IP for node %s: %s", node.Name, err)
+		return 0, err
+	}
+
+	// Get value from cache if exists
+	client.cacheLock.Lock()
+	entry, entryPresent := client.cache[ip]
+	if entryPresent {
+		if entry.nodeCpuFreq != nil {
+			nodeFreq := *entry.nodeCpuFreq
+			client.cacheLock.Unlock()
+			return nodeFreq, nil
+		}
+	}
+	client.cacheLock.Unlock()
+
+	// We could not get a valid cpu frequency from cache, discover it.
+	nodeFreq := client.defaultCpuFreq
+	minfo, err := client.GetMachineInfo(ip)
+	if err != nil {
+		glog.Errorf("Failed to get machine[%s] cpu.frequency from kubelet: %v. Will try pod based getter.", node.Name, err)
+		nodeIsActive := util.NodeIsReady(node) && util.NodeIsSchedulable(node)
+		if nodeIsActive {
+			nodeFreq, err = client.fallbkCpuFreqGetter.GetFrequency(node.Name)
+			if err != nil {
+				glog.Errorf("Failed to get cpufreq from getter %s: %s. Default value will be used.", node.Name, err)
+			}
```

Comment:
maybe you should also do client.defaultCpuFreq = nodeFreq in case jobs are created successfully for some of the nodes, we may want to use this value for the failed ones.

Reply:
Its exactly what is done on line 209 below.
So the default cpuFreq with be updated even if it could be discovered from only one node.

Reply:
Line 209 is only invoked when machineinfo is available. My point is if we fall back to the job approach, and only some of the jobs are successfully deployed (others are not schedulable for example), then we should still override the default with the one we got from the successful job.

Reply:
You are right. Will update.